src/internal/internal_gerbt.cc

```diff
@@ -0,0 +1,470 @@
+// Copyright (c) 2020-2022, University of Tennessee. All rights reserved.
+// SPDX-License-Identifier: BSD-3-Clause
+// This program is free software: you can redistribute it and/or modify it under
+// the terms of the BSD 3-Clause license. See the accompanying LICENSE file.
+
+#include "slate/types.hh"
+#include "internal/internal.hh"
+#include "internal/Tile_gerbt.hh"
+
+
+namespace slate {
+
+namespace internal {
+
+template<typename scalar_t>
+void gerbt(Matrix<scalar_t> A11,
+           Matrix<scalar_t> A12,
+           Matrix<scalar_t> A21,
+           Matrix<scalar_t> A22,
+           Matrix<scalar_t> U1,
+           Matrix<scalar_t> U2,
+           Matrix<scalar_t> V1,
+           Matrix<scalar_t> V2)
+{
+    // Assuming U and V have same structure as the A's where they're multiplied
+
+    slate_assert(A11.mt() >= A22.mt());
+    slate_assert(A11.nt() >= A22.nt());
+
+    slate_assert(A11.mt() == A12.mt());
+    slate_assert(A22.nt() == A12.nt());
+    slate_assert(A22.mt() == A21.mt());
+    slate_assert(A11.nt() == A21.nt());
+
+    const int64_t mt = A22.mt();
+    const int64_t nt = A22.nt();
+    const int64_t mt_full = A11.mt();
+    const int64_t nt_full = A11.nt();
+
+    // Used to manage OpenMP task dependencies
+    std::vector<uint8_t> task_vect (mt_full*nt_full);
+    uint8_t* task = task_vect.data();
+    SLATE_UNUSED( task ); // Used only by OpenMP
+
+    std::vector<MPI_Request> requests;
+    for (int64_t ii = 0; ii < mt_full; ++ii) {
+        for (int64_t jj = 0; jj < nt_full; ++jj) {
+            const int64_t tag = 4*(ii*nt_full + jj);
+            MPI_Request r;
+            if (A11.tileIsLocal(ii, jj)) {
+                if (jj < nt) {
+                    A12.tileIrecv( ii, jj, A12.tileRank(ii, jj),
+                                  Layout::ColMajor, tag+1, &r );
+                    if (r != MPI_REQUEST_NULL) {
```

Comment:
What happens if `r == MPI_REQUEST_NULL`?

Reply:
There's nothing we need to wait on (specifically it's a local tile).  So, there's no point in expanding our vector of requests with no-ops.